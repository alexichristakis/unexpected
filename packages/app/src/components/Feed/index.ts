export { default as Posts } from "./Posts";
export * from "./Top";
