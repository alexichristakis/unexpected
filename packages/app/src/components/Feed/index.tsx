import React, { useCallback, useState, useRef, useEffect } from "react";
import { View, StyleSheet, Text, ScrollView } from "react-native";
import Animated, {
  useCode,
  debug,
  Value,
  Clock,
} from "react-native-reanimated";
import { connect, ConnectedProps } from "react-redux";
import { PanGestureHandler, State } from "react-native-gesture-handler";
import {
  useGestureHandler,
  useValue,
  useValues,
  //   withSpring,
  WithSpringParams,
  snapPoint,
  useClock,
  withSpring,
  min,
} from "react-native-redash";
import { useMemoOne } from "use-memo-one";

import * as selectors from "@redux/selectors";
import { RootState } from "@redux/types";
import { SCREEN_HEIGHT, SCREEN_WIDTH, SPRING_CONFIG, Colors } from "@lib";
import Swipeable from "./Swipeable";
import Post, { POST_HEIGHT } from "@components/Post";
import random from "lodash/random";
import { TextStyles } from "@lib";

import Header from "./Header";

const {
  onChange,
  cond,
  abs,
  modulo,
  divide,
  round,
  multiply,
  set,
  and,
  clockRunning,
  not,
  neq,
  call,
  block,
  startClock,
  stopClock,
  spring,
  sub,
  add,
  eq,
  lessThan,
  greaterThan,
} = Animated;

const connector = connect(
  (state: RootState) => ({ postIds: selectors.allPosts(state) }),
  {}
);

export interface FeedProps {}

export type FeedConnectedProps = ConnectedProps<typeof connector>;

const Feed: React.FC<FeedProps & FeedConnectedProps> = ({ postIds }) => {
  const state = useValue(State.UNDETERMINED);
  const [value, velocity, index] = useValues<number>([0, 0, 0]);

  const handler = useGestureHandler({
    state,
    translationY: value,
    velocityY: velocity,
  });

  const translateY = useMemoOne(() => {
    const clock = new Clock();
    const offset = new Value(0);
    const springState: Animated.SpringState = {
      finished: new Value(0),
      velocity: new Value(0),
      position: new Value(0),
      time: new Value(0),
    };

    const config = {
      toValue: new Value(0),
      ...SPRING_CONFIG,
    };

    const gestureAndAnimationIsOver = new Value(1);
    const isSpringInterrupted = and(
      eq(state, State.BEGAN),
      clockRunning(clock)
    );

    const finishSpring = [
      set(offset, springState.position),
      stopClock(clock),
      set(gestureAndAnimationIsOver, 1),
    ];

    const next = cond(
      neq(index, postIds.length - 1),
      sub(offset, POST_HEIGHT),
      0
    );
    const prev = cond(neq(index, 0), add(offset, POST_HEIGHT), 0);

    const point = snapPoint(springState.position, velocity, [
      next,
      offset,
      prev,
    ]);

    return block([
      cond(isSpringInterrupted, finishSpring),
      cond(gestureAndAnimationIsOver, set(springState.position, offset)),
      cond(neq(state, State.END), [
        set(gestureAndAnimationIsOver, 0),
        set(springState.finished, 0),
        set(springState.position, add(offset, value)),
      ]),
      cond(and(eq(state, State.END), not(gestureAndAnimationIsOver)), [
        cond(and(not(clockRunning(clock)), not(springState.finished)), [
          set(springState.velocity, velocity),
          set(springState.time, 0),
          set(config.toValue, point),
          startClock(clock),
        ]),
        spring(clock, springState, config),
        cond(clockRunning(clock), [
          set(index, round(divide(springState.position, -POST_HEIGHT))),

          //   springState.position
          // snap complete
        ]),
        // onChange(index, [
        //   //
        //   //
        // ]),
        cond(springState.finished, [
          //   cond(clockRunning(clock), [
          //     set(index, round(divide(springState.position, -POST_HEIGHT))),

          //     //   springState.position
          //     // snap complete
          //   ]),
          ...finishSpring,
        ]),
      ]),
      springState.position,
    ]);
  }, []);

  return (
    <View style={styles.container}>
      <PanGestureHandler activeOffsetY={[-10, 10]} {...handler}>
        <Animated.View style={{ transform: [{ translateY }] }}>
          {postIds.map((id, key) => (
            <Post
              visible={eq(index, key)}
              dragStarted={eq(state, State.BEGAN)}
              offset={sub(translateY, key * -POST_HEIGHT)}
              {...{ id, key }}
            />
          ))}
        </Animated.View>
      </PanGestureHandler>
      <Header offset={translateY} />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    overflow: "hidden",
    backgroundColor: Colors.background,
    paddingTop: 150,
  },
});

export default connector(Feed);
