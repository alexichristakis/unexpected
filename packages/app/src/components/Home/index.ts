export * from "./Pager";
export * from "./Header";
export * from "./TabBar";
export * from "./ReactiveOverlay";
