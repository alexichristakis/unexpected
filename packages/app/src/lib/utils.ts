import Animated, { Easing } from "react-native-reanimated";
import { User } from "@unexpected/global";

export const formatName = (user?: User) =>
  user?.firstName && user?.lastName ? `${user.firstName} ${user.lastName}` : "";
