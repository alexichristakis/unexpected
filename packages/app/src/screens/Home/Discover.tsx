import React, { useState } from "react";
import {
  ActivityIndicator,
  FlatList,
  ListRenderItemInfo,
  NativeSyntheticEvent,
  StyleSheet,
  Text,
  TextInputSubmitEditingEventData,
  View
} from "react-native";

import client, { getHeaders } from "@api";
import { RouteProp } from "@react-navigation/core";
import { NativeStackNavigationProp } from "@react-navigation/native-stack";
import { Screen } from "react-native-screens";
import { connect, ConnectedProps } from "react-redux";
import uuid from "uuid/v4";

import { Input, ItemSeparator, UserRow } from "@components/universal";
import { useDarkStatusBar } from "@hooks";
import { SB_HEIGHT, TextSizes, TextStyles } from "@lib/styles";
import { Actions as UserActions } from "@redux/modules/user";
import * as selectors from "@redux/selectors";
import { RootState } from "@redux/types";
import { User } from "@unexpected/global";

import { ParamList } from "../../App";

const mapStateToProps = (state: RootState) => ({
  phoneNumber: selectors.phoneNumber(state),
  jwt: selectors.jwt(state)
});
const mapDispatchToProps = {
  loadUsers: UserActions.loadUsers
};

export type DiscoverConnectorProps = ConnectedProps<typeof connector>;
export interface DiscoverProps {
  navigation: NativeStackNavigationProp<ParamList, "DISCOVER">;
  route: RouteProp<ParamList, "DISCOVER">;
}
export const Discover: React.FC<DiscoverProps &
  DiscoverConnectorProps> = React.memo(
  ({ phoneNumber, jwt, navigation, loadUsers }) => {
    const [responses, setResponses] = useState<User[]>([]);
    const [loading, setLoading] = useState(false);

    useDarkStatusBar();

    const renderUserRow = ({ item, index }: ListRenderItemInfo<User>) => (
      <UserRow onPress={handleOnPressUser} user={item} />
    );

    const handleOnPressUser = (toUser: User) => {
      if (phoneNumber === toUser.phoneNumber) {
        navigation.navigate("USER_PROFILE_TAB");
      } else {
        navigation.navigate({
          name: "PROFILE",
          key: uuid(),
          params: {
            prevRoute: "Search",
            phoneNumber: toUser.phoneNumber
          }
        });
      }
    };

    const handleOnPressSubmit = async ({
      nativeEvent
    }: NativeSyntheticEvent<TextInputSubmitEditingEventData>) => {
      setLoading(true);

      const response = await client.get<User[]>(
        `/user/search/${nativeEvent.text}`,
        {
          headers: getHeaders({ jwt })
        }
      );

      const { data } = response;

      const results = data.length
        ? data.filter(o => o.phoneNumber !== phoneNumber)
        : [];

      loadUsers(results);
      setResponses(results);
      setLoading(false);
    };

    const renderEmptyComponent = () => (
      <View style={{ padding: 20 }}>
        <Text style={TextStyles.medium}>No results</Text>
      </View>
    );

    const renderSeparatorComponent = () => <ItemSeparator />;

    return (
      <Screen style={styles.container}>
        <Input
          size={TextSizes.title}
          style={{ width: "100%", paddingHorizontal: 20 }}
          returnKeyType={"search"}
          label="enter a name or phone number"
          placeholder="search"
          onSubmitEditing={handleOnPressSubmit}
        />
        {loading ? (
          <ActivityIndicator size="large" />
        ) : (
          <FlatList
            style={styles.list}
            renderItem={renderUserRow}
            ListEmptyComponent={renderEmptyComponent}
            ItemSeparatorComponent={renderSeparatorComponent}
            data={responses}
          />
        )}
      </Screen>
    );
  }
);

const styles = StyleSheet.create({
  container: {
    paddingTop: SB_HEIGHT,
    // paddingHorizontal: 20,
    alignItems: "center"
  },
  list: { height: "100%", width: "100%" }
});

const connector = connect(mapStateToProps, mapDispatchToProps);
export default connector(Discover);
