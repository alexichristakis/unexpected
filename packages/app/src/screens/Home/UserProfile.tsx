import React, { useCallback, useState } from "react";
import {
  NativeScrollEvent,
  NativeSyntheticEvent,
  StatusBar,
  StyleSheet
} from "react-native";

import { RouteProp, useFocusEffect } from "@react-navigation/core";
import { NativeStackNavigationProp } from "@react-navigation/native-stack";
import isEqual from "lodash/isEqual";
import Haptics from "react-native-haptic-feedback";
import Animated from "react-native-reanimated";
import { Screen } from "react-native-screens";
import { connect } from "react-redux";
import uuid from "uuid/v4";

import { Top } from "@components/Profile";
import { Grid } from "@components/Profile/Grid";
import { SB_HEIGHT } from "@lib/styles";
import { Actions as AuthActions } from "@redux/modules/auth";
import { Actions as PostActions } from "@redux/modules/post";
import { Actions as UserActions } from "@redux/modules/user";
import * as selectors from "@redux/selectors";
import { ReduxPropsType, RootState } from "@redux/types";
import { Post } from "@unexpected/global";
import { StackParamList } from "../../App";

const mapStateToProps = (state: RootState) => ({
  postsLoading: selectors.postLoading(state),
  user: selectors.currentUser(state),
  posts: selectors.currentUsersPosts(state),
  stale: selectors.currentUsersPostsStale(state)
});
const mapDispatchToProps = {
  logout: AuthActions.logout,
  fetchUser: UserActions.fetchUser,
  fetchUsersPosts: PostActions.fetchUsersPosts
};

export type UserProfileReduxProps = ReduxPropsType<
  typeof mapStateToProps,
  typeof mapDispatchToProps
>;
export interface UserProfileOwnProps {
  navigation: NativeStackNavigationProp<StackParamList, "USER_PROFILE">;
  route: RouteProp<StackParamList, "USER_PROFILE">;
}
export type UserProfileProps = UserProfileOwnProps & UserProfileReduxProps;

export const UserProfile: React.FC<UserProfileProps> = React.memo(
  ({
    navigation,
    fetchUser,
    fetchUsersPosts,
    stale,
    posts,
    postsLoading,
    user
  }) => {
    const [scrollY] = useState(new Animated.Value(0));

    useFocusEffect(
      useCallback(() => {
        StatusBar.setHidden(false);
        fetchUser();

        if (stale) fetchUsersPosts();

        return () => {};
      }, [stale])
    );

    const goToNewProfilePicture = () => {
      navigation.navigate("NEW_PROFILE_PICTURE");
    };

    const goToSettings = () => {
      navigation.navigate("SETTINGS");
    };

    const goToFriends = () => {
      navigation.navigate("FRIENDS", { user });
    };

    const goToEditBio = () => {
      navigation.navigate("EDIT_BIO");
    };

    const renderTop = () => (
      <Top
        isUser={true}
        user={user}
        numPosts={posts.length}
        scrollY={scrollY}
        onPressAddBio={goToEditBio}
        onPressFriends={goToFriends}
        onPressImage={goToNewProfilePicture}
        onPressName={goToSettings}
      />
    );

    const handleOnPressPost = (post: Post) => {
      navigation.navigate({
        name: "POST",
        key: uuid(),
        params: { prevRoute: user.firstName, post: { ...post, user } }
      });
    };

    const handleOnScrollEndDrag = (
      event: NativeSyntheticEvent<NativeScrollEvent>
    ) => {
      const {
        nativeEvent: {
          contentOffset: { y }
        }
      } = event;

      if (y < -100) {
        Haptics.trigger("impactMedium");
        fetchUser();
        fetchUsersPosts();
      }
    };

    return (
      <Screen style={styles.container}>
        <Grid
          loading={postsLoading}
          onPressPost={handleOnPressPost}
          scrollY={scrollY}
          onScrollEndDrag={handleOnScrollEndDrag}
          ListHeaderComponentStyle={styles.headerContainer}
          ListHeaderComponent={renderTop}
          posts={posts}
        />
      </Screen>
    );
  },
  (prevProps, nextProps) =>
    prevProps.stale === nextProps.stale &&
    isEqual(prevProps.user, nextProps.user)
);

const styles = StyleSheet.create({
  container: {
    paddingTop: SB_HEIGHT(),
    alignItems: "center"
  },
  headerContainer: {
    zIndex: 1,
    alignItems: "center",
    alignSelf: "stretch"
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(UserProfile);
