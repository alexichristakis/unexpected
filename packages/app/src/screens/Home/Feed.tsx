import React, { useCallback, useEffect, useState, useRef } from "react";
import {
  NativeScrollEvent,
  NativeSyntheticEvent,
  StatusBar,
  StyleSheet,
  FlatList
} from "react-native";

import { useScrollToTop } from "@react-navigation/native";
import { RouteProp, useFocusEffect } from "@react-navigation/core";
import { NativeStackNavigationProp } from "@react-navigation/native-stack";
import _ from "lodash";
import Haptics from "react-native-haptic-feedback";
import Animated, { Easing } from "react-native-reanimated";
import { Screen } from "react-native-screens";
import { connect } from "react-redux";
import uuid from "uuid/v4";

import { Posts } from "@components/Feed";
import { ZoomedImage, ZoomedImageType } from "@components/universal";
import { SB_HEIGHT } from "@lib/styles";
import { Actions as PostActions } from "@redux/modules/post";
import * as selectors from "@redux/selectors";
import { ReduxPropsType, RootState } from "@redux/types";

import { StackParamList } from "../../App";
import { hideStatusBarOnScroll } from "@hooks";

const { Value, block, cond, call, greaterOrEq, useCode } = Animated;

const mapStateToProps = (state: RootState) => ({
  stale: selectors.feedStale(state),
  phoneNumber: selectors.phoneNumber(state),
  refreshing: selectors.feedLoading(state),
  shouldLaunchPermissions: selectors.shouldLaunchPermissions(state)
});
const mapDispatchToProps = {
  fetchFeed: PostActions.fetchFeed
};

export type FeedReduxProps = ReduxPropsType<
  typeof mapStateToProps,
  typeof mapDispatchToProps
>;
export interface FeedProps extends FeedReduxProps {
  navigation: NativeStackNavigationProp<StackParamList, "FEED">;
  route: RouteProp<StackParamList, "FEED">;
}

export const Feed: React.FC<FeedProps> = React.memo(
  ({
    stale,
    navigation,
    phoneNumber,
    fetchFeed,
    refreshing,
    shouldLaunchPermissions
  }) => {
    const [zoomedImage, setZoomedImage] = useState<ZoomedImageType>();
    const [scrollY] = useState(new Value(0));
    const scrollRef = useRef<FlatList>(null);

    const animatedStatusBarStyle = hideStatusBarOnScroll(scrollY);

    // @ts-ignore
    useScrollToTop(scrollRef);

    useEffect(() => {
      // fetchFeed();

      if (shouldLaunchPermissions) {
        setTimeout(() => navigation.navigate("PERMISSIONS"), 100);
      }
    }, []);

    const handleOnScrollEndDrag = (
      event: NativeSyntheticEvent<NativeScrollEvent>
    ) => {
      const {
        nativeEvent: {
          contentOffset: { y }
        }
      } = event;

      if (y < -100) {
        Haptics.trigger("impactMedium");
        fetchFeed();
      }
    };

    const handleOnPressUser = (userPhoneNumber: string) => {
      if (phoneNumber === userPhoneNumber) {
        navigation.navigate("USER_PROFILE");
      } else {
        navigation.navigate({
          name: "PROFILE",
          key: uuid(),
          params: { prevRoute: "Feed", phoneNumber: userPhoneNumber }
        });
      }
    };

    const handleOnPressShare = () => {
      navigation.navigate("CAPTURE");
    };

    const handleOnGestureComplete = () => setZoomedImage(undefined);

    return (
      <Screen style={styles.container}>
        <Posts
          scrollRef={scrollRef}
          scrollY={scrollY}
          refreshing={refreshing}
          onScrollEndDrag={handleOnScrollEndDrag}
          onGestureBegan={setZoomedImage}
          onGestureComplete={handleOnGestureComplete}
          onPressUser={handleOnPressUser}
          onPressShare={handleOnPressShare}
        />
        {zoomedImage && <ZoomedImage {...zoomedImage} />}
        <Animated.View style={[styles.statusBar, animatedStatusBarStyle]} />
      </Screen>
    );
  }
);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    // paddingHorizontal: 20,
    alignItems: "center"
  },
  statusBar: {
    position: "absolute",
    left: 0,
    right: 0,
    top: 0,
    height: SB_HEIGHT(),
    backgroundColor: "white"
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(Feed);
