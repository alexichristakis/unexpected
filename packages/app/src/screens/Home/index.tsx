import React from "react";
import { StyleSheet, Text, View } from "react-native";
import { NativeStackNavigationProp } from "react-native-screens/native-stack";
import { connect, ConnectedProps } from "react-redux";
import { useVector } from "react-native-redash";
import Animated from "react-native-reanimated";

import { Pager, TabBar } from "@components/Home";

import { RootState as RootStateType } from "@redux/types";
import Feed from "@components/Feed";
import Profile from "@components/Profile";
import { Colors } from "@lib";

import { StackParamList } from "../../App";

const connector = connect((state: RootStateType) => ({}), {});

export type HomeReduxProps = ConnectedProps<typeof connector>;
export interface HomeOwnProps {
  navigation: NativeStackNavigationProp<StackParamList>;
}

const Home: React.FC<HomeReduxProps & HomeOwnProps> = ({ navigation }) => {
  const offset = useVector(0, 0);

  return (
    <View style={{ flex: 1, backgroundColor: Colors.gray }}>
      <Animated.View
        style={{
          flex: 1,
          transform: [{ translateY: offset.y }],
        }}
      >
        <Pager {...offset}>
          <Feed />
          <Profile />
        </Pager>
        <TabBar {...offset} />
      </Animated.View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
});

export default connector(Home);
