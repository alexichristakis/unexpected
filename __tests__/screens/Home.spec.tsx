import "react-native";
import React from "react";
import { mount } from "enzyme";
// import renderer from "react-test-renderer";

import { useNavigation, useContext } from "../utils";
import { Feed, FeedProps as FeedPropsType } from "@screens/Home/Feed";
import { UserProfile, UserProfileProps as UserProfilePropsType } from "@screens/Home/UserProfile";
import { Discover, DiscoverProps as DiscoverPropsType } from "@screens/Home/Discover";

describe("home", () => {
  const feedProps: FeedPropsType = {
    uploadPhoto: jest.fn(),
    logout: jest.fn(),
    requestNotificationPermissions: jest.fn(),
    requestPermission: jest.fn()
  };
  const userProfileProps: UserProfilePropsType = {
    logout: jest.fn(),
    fetchUsersPosts: jest.fn(),
    posts: [],
    stale: false,
    user: {
      firstName: "Test",
      lastName: "Person",
      deviceOS: "",
      deviceToken: "",
      timezone: "",
      following: [],
      phoneNumber: ""
    }
  };
  const discoverProps: DiscoverPropsType = {};

  //   const feed = mount(useContext(useNavigation(<Feed {...feedProps} />)));
  const profile = mount(useContext(useNavigation(<UserProfile {...userProfileProps} />)));
  //   const discover = mount(useContext(<Discover {...discoverProps} />));

  describe("profile", () => {
    it("renders correctly", async () => {
      expect(profile).toMatchSnapshot();
    });
  });

  //   describe("feed", () => {
  //     it("renders correctly", async () => {
  //       expect(feed).toMatchSnapshot();
  //     });
  //   });

  //   describe("discover", () => {
  //     it("renders correctly", async () => {
  //       expect(discover).toMatchSnapshot();
  //     });
  //   });
});
