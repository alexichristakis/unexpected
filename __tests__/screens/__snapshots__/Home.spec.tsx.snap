// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`home profile renders correctly 1`] = `
<Context>
  <Provider
    store={
      Object {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <SafeAreaProvider>
      <RNCSafeAreaView
        onInsetsChange={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <RNCSafeAreaView
          onInsetsChange={[Function]}
          style={
            Object {
              "flex": 1,
            }
          }
        />
      </RNCSafeAreaView>
    </SafeAreaProvider>
  </Provider>
</Context>
`;
